dulwich/porcelain.py

```diff
@@ -1859,6 +1864,107 @@ def reset_file(repo, file_path: str, target: bytes = b'HEAD',
     build_file_from_blob(blob, mode, full_path, symlink_fn=symlink_fn)
 
 
+def _update_head_during_checkout_branch(repo, target):
+    checkout_target = None
+    if target == b'HEAD':  # Do not update head while trying to checkout to HEAD.
+        pass
+    elif target in repo.refs.keys(base=LOCAL_BRANCH_PREFIX):
+        update_head(repo, target)
+    else:
+        # If checking out a remote branch, create a local one without the remote name prefix.
+        config = repo.get_config()
+        name = target.split(b"/")[0]
+        section = (b"remote", name)
+        if config.has_section(section):
+            checkout_target = target.replace(name + b"/", b"")
+            try:
+                branch_create(repo, checkout_target, (LOCAL_REMOTE_PREFIX + target).decode())
+            except Error:
+                pass
+            update_head(repo, LOCAL_BRANCH_PREFIX + checkout_target)
+        else:
+            update_head(repo, target, detached=True)
+
+    return checkout_target
+
+
+def checkout_branch(repo, target: Union[bytes, str], force: bool = False):
+    """switch branches or restore working tree files.
+
+    The implementation of this function will probably not scale well
+    for branches with lots of local changes.
+    This is due to the analysis of a diff between branches before any
+    changes are applied.
+
+    Args:
+      repo: dulwich Repo object
+      target: branch name or commit sha to checkout
+      force: true or not to force checkout
+    """
+    target = to_bytes(target)
+
+    current_tree = parse_tree(repo, repo.head())
+    target_tree = parse_tree(repo, target)
+
+    if force:
+        repo.reset_index(target_tree.id)
+        _update_head_during_checkout_branch(repo, target)
+    else:
+        status_report = status(repo)
+        changes = list(set(status_report[0]['add'] + status_report[0]['delete'] + status_report[0]['modify'] + status_report[1]))
+        index = 0
+        while index < len(changes):
+            change = changes[index]
+            try:
+                current_tree.lookup_path(repo.object_store.__getitem__, change)
+                try:
+                    target_tree.lookup_path(repo.object_store.__getitem__, change)
+                    index += 1
+                except KeyError:
+                    raise AbortCheckout('Your local changes to the following files would be overwritten by checkout: ' + change.decode())
+            except KeyError:
+                changes.pop(index)
+
+        # Update head.
+        checkout_target = _update_head_during_checkout_branch(repo, target)
+        if checkout_target is not None:
+            target_tree = parse_tree(repo, checkout_target)
+
+        dealt_with = []
```

Comment:
could this be a set, for better performance?

Reply:
[daaf927](https://github.com/jelmer/dulwich/pull/1148/commits/daaf92711c725e989d670761a961ffd2ab3d59e5)